### ISSUE TITLE
Intermittent leveldb mapreduce failure: skip/limit and multiple keys/values

### ISSUE THREAD

**Issue by nolanlawson** [pouchdb/pouchdb#3082]
Seeing this in some failed mapreduce builds:

```
local with temp views: views should query correctly with skip/limit and multiple keys/values:

AssertionError: correctly return rows: expected [ Array(2) ] to have a length of 1 but got 2
```

**Comment by slang800** [comment]
I'm getting something similar... I have a db with 2715 docs (using leveldb as a backend), and I'm requesting them in batches of 100 (to avoid loading everything into memory at once). But sometimes I get more than 100 docs back. Here's a list of results from incrementally requesting the database in chunks:

```
got 101 docs
got 102 docs
got 100 docs
got 100 docs
got 101 docs
got 100 docs
got 100 docs
got 103 docs
got 100 docs
got 100 docs
got 101 docs
got 100 docs
got 100 docs
got 100 docs
got 102 docs
got 100 docs
got 100 docs
got 100 docs
got 100 docs
got 100 docs
got 101 docs
got 102 docs
got 100 docs
got 101 docs
got 100 docs
got 100 docs
got 100 docs
got 1 docs
```

it returns the correct number of docs, but it is exceeding the limit I set

**Comment by nolanlawson** [comment]
Can you write a test to reproduce? Also do you see this error only in LevelDB or also in IndexedDB/WebSQL?

**Comment by slang800** [comment]
This is running only on nodejs, so I don't think I can use it with IndexedDB/WebSQL, right?

**Comment by slang800** [comment]
I did test a direct connection to the main server (using the http adapter), and that gives no issues with being over the limit, and doesn't skip any docs. So I think this is a problem in the LevelDB adapter

**Comment by slang800** [comment]
Oh, and it seems like `total_rows` in the response is correct for leveldb (because it matches what couchdb's /_utils reports) - it's somehow skipping rows.

**Comment by nolanlawson** [comment]
Looks like we may have a race condition in the LevelDB adapter that allows it to return more results than requested. Again, a reproducible test case would be excellent here, because otherwise it's really hard to tell what's going on. :)

**Comment by slang800** [comment]
I haven't been able to reproduce it consistently enough to make a test-case for it... It happens with like 20% of all requests, and that's with me requesting 100 docs each time. Maybe if we just made the test case that you showed above request a lot more docs? If I figure out how to reproduce it more reliably, I'll be sure to make a case, but at this point my only lead is "request a lot of data in big chunks".

**Comment by nolanlawson** [comment]
Hm, when you call `allDocs()`, are you using `startkey` or `endkey` or any of those parameters? When you get the extra doc, is it one too many from the end of the list or from the start of the list?

**Comment by nolanlawson** [comment]
Also, can you try this branch and let me know if it fixes it for you? https://github.com/pouchdb/pouchdb/tree/3082 Just add to your `package.json`:

```
"pouchdb": "pouchdb/pouchdb#3082"
```

If it fixes the issue, then I'll try to write a test to reproduce it. Or if you would be so enterprising as to add a test to `test.all_docs.js`... :wink:

**Comment by slang800** [comment]
I'm calling with only `startkey`, `limit`, and `include_docs` defined. This is the actual code that I'm using:
```coffee
Readable = require('readable-stream').Readable
map = require 'map-stream'

###*
 * Kinda returns a stream of all the docs in the DB. It's actually made by
   paginating through the DB and pushing that into a stream, but it's close
   enough.
###
sortaStreamAllDocs = (db, includeDocs = false) ->
  output = new Readable(objectMode: true)
  lastKey = null
  lock = false
  uniqueKeys = []

  write = (rows...) ->
    for row in rows
      if row is null
        output.push null
      else
        if row.id in uniqueKeys
          throw new Error('non-unique key ' + row.id)
        uniqueKeys.push(row.id)

        output.push (
          if includeDocs
            row.doc
          else
            {_id: row.id}
        )

  # we're actually making an assumption that the database isn't gaining/losing
  # any docs while this function is running.
  outputSoFar = 0

  output._read = ->
    # prevent additional `allDocs` queries from being started while one is
    # already running
    if lock then return

    lock = true

    options =
      limit: 100
      include_docs: includeDocs

    if lastKey isnt null
      options.startkey = lastKey

    db.allDocs(options).then((result) ->
      if result.rows.length > options.limit
        log.warn(
          "utils/sortaStreamAllDocs: docs returned > limit"
          docsReturned: result.rows.length
        )

      if lastKey isnt null
        # we have to discard the first key, because we already output it in the
        # last request... if we could set startKey to be non-inclusive, or if we
        # could determine what the next key is then we could avoid this
        result.rows.shift()

      outputSoFar += result.rows.length

      if result.rows.length is 0
        if outputSoFar isnt result['total_rows']
          console.warn "missing #{result['total_rows'] - outputSoFar} rows"
        write(null)
        return

      lastRow = result.rows.pop()
      write(result.rows...)

      if outputSoFar is result['total_rows']
        # done normally
        console.warn "ending normally #{outputSoFar} / #{result['total_rows']}"
        write(lastRow, null)
      else if outputSoFar > result['total_rows']
        console.error('got too many rows?!')
        write(lastRow, null)
      else
        lastKey = lastRow.id
        lock = false # need to unlock before sending the last row
        console.warn "still going... #{outputSoFar} / #{result['total_rows']}"
        write(lastRow)
    )
    return

  return output.pipe(map((doc, cb) ->
    # skip design docs
    if doc._id[0] is '_' then cb(null) else cb(null, doc)
  ))
```
...there's some extra logging in there just from debugging without a proper debugger, and `log.warn` is from winston, for sending the issue with going over the limit to a graylog instance for recording.

And sure, I'll test it out with that branch.

**Comment by slang800** [comment]
Yeah, that definitely didn't work... `limit` gets mutated somewhere in that function, so by the time it gets to the check you added it's like `10`, `19`, `37`, or `40` ... among other values I saw.

**Comment by nolanlawson** [comment]
oh right, forgot about that. okay, I'll try another fix

**Comment by nolanlawson** [comment]
okay, try that branch again

**Comment by nolanlawson** [comment]
Also, I wrote a [test](https://github.com/pouchdb/pouchdb/blob/6458f5e7431e5b8a888ce8db451ba6e44ac03a1b/tests/integration/test.all_docs.js#L584-L607) to try to reproduce, but I just cannot. What version of LevelDOWN do you have installed, and on what architecture?

**Comment by nolanlawson** [comment]
Also I notice that you're not adding `{skip: 1}` to your options anywhere. `startkey` is inclusive, so it will definitely include that doc at the start of the list.

**Comment by daleharvey** [comment]
Hopefully fixed by https://github.com/pouchdb/pouchdb/commit/b59a2c12383091ac35146f1cddee70adbe180a28

**Comment by slang800** [comment]
Yep, I think that fixed it! Finally upgraded to v3.5 & I'm not seeing the error on any servers.
